### PR TITLE
Martin/ios 482 deeplink from message

### PR DESCRIPTION
**Pull request by MartinJibber**
https://user-images.githubusercontent.com/88675954/155242289-4169b684-73ca-4930-8349-8ebb17bfda9c.mp4

### CODE REVIEW

File: Shared/Coordinator/MainCoordinator.swift [hunk 1]
```diff
@@ -32,10 +32,23 @@ class MainCoordinator: Coordinator<Void> {
         self.runLaunchFlow()
     }
 
+    /// A task to start the launch flow and handle a deep link. If the task is cancelled, the deep link will not be handled.
+    private var launchAndDeepLinkTask: Task<Void, Never>?
```
Comment (MartinJibber) [inline Shared/Coordinator/MainCoordinator.swift:36]: A task let's us keep track of whether we're still launching.

File: Shared/Coordinator/MainCoordinator.swift [hunk 2]
```diff
@@ -32,10 +32,23 @@ class MainCoordinator: Coordinator<Void> {
...
+                let launchCoordinator = LaunchCoordinator(router: self.router, deepLink: self.deepLink)
+                self.router.setRootModule(launchCoordinator)
+                self.addChildAndStart(launchCoordinator) { launchStatus in
+                    continuation.resume(returning: launchStatus)
+                }
+            }
+
+            // Don't handle the launch status if the task was cancelled.
+            guard !Task.isCancelled else { return }
+
```
Comment (MartinJibber) [inline Shared/Coordinator/MainCoordinator.swift:51]: The whole launch flow is wrapped in a task.

File: Shared/Coordinator/MainCoordinator.swift [hunk 3]
```diff
@@ -147,10 +160,19 @@ class MainCoordinator: Coordinator<Void> {
...
     nonisolated func userNotificationManager(willHandle deeplink: DeepLinkable) {
-        Task.onMainActor {
+        Task.onMainActorAsync {
+            // Cancelling other deeplink calls.
+            self.launchAndDeepLinkTask?.cancel()
+
+            // Wait until the launch is finished.
+            await self.launchAndDeepLinkTask?.value
+
             self.handle(deeplink: deeplink)
```
Comment (MartinJibber) [inline Shared/Coordinator/MainCoordinator.swift:176]: Now we'll only handle the deep link once we're finished. Cancelling ensures that we don't attempt to deep link twice.

File: Shared/Extensions/Task+Extensions.swift [hunk 4]
```diff
@@ -20,9 +20,9 @@ extension Task where Success == Void, Failure == Never {
     }
 
     /// Creates a new task that runs the passed in closure on the MainActor. For use in async functions.
-    static func onMainActorAsync(body: @escaping @MainActor @Sendable () -> Success) async {
-        await MainActor.run {
-            body()
+    static func onMainActorAsync(body: @escaping @MainActor @Sendable () async -> Success) {
+        Task {
+            await body()
```
Comment (MartinJibber) [inline Shared/Extensions/Task+Extensions.swift:25]: Made this function much more useful. It's not async anymore, but you pass in an async function that's guaranteed to run on main.

File: Shared/Flows/Launch/LaunchCoordinator.swift [hunk 5]
```diff
@@ -20,7 +20,7 @@ class LaunchCoordinator: PresentableCoordinator<LaunchStatus> {
         super.start()
 
         Task {
-            let launchStatus = await LaunchManager.shared.launchApp()
+            let launchStatus = await LaunchManager.shared.launchApp(with: self.deepLink)
```
Comment (MartinJibber) [inline Shared/Flows/Launch/LaunchCoordinator.swift:23]: Doesn't really make a difference now, but now the launch manager actually uses the passed in deep link.

File: Shared/Notifications/UserNotificationManager.swift [hunk 6]
```diff
@@ -26,8 +26,6 @@ class UserNotificationManager: NSObject {
     private let center = UNUserNotificationCenter.current()
     private(set) var application: UIApplication?
 
-    var cancellables = Set<AnyCancellable>()
```
Comment (MartinJibber) [inline Shared/Notifications/UserNotificationManager.swift:29]: Unused.

File: Shared/Notifications/UserNotificationManager.swift [hunk 7]
```diff
@@ -95,7 +92,24 @@ class UserNotificationManager: NSObject {
             return false
         }
     }
-    
+
```
Comment (MartinJibber) [inline Shared/Notifications/UserNotificationManager.swift:95]: Reorganization.

File: Shared/Notifications/UserNotificationManager.swift [hunk 8]
```diff
@@ -116,35 +130,61 @@ class UserNotificationManager: NSObject {
...
     }
-    
-    func getDeliveredNotifications() async -> [UNNotification] {
-        return await withCheckedContinuation({ continuation in
-            self.center.getDeliveredNotifications { notifications in
-                continuation.resume(returning: notifications)
+}
+
+// MARK: - UNUserNotificationCenterDelegate
+
```
Comment (MartinJibber) [inline Shared/Notifications/UserNotificationManager.swift:150]: Reorganization.

File: Shared/Notifications/UserNotificationManager.swift [hunk 9]
```diff
@@ -116,49 +144,61 @@ class UserNotificationManager: NSObject {
...
-    }
 
-    func registerPush(from deviceToken: Data) async {
-        do {
-            let installation = try await PFInstallation.getCurrent()
-            installation.badge = 0
-            installation.setDeviceTokenFrom(deviceToken)
+    func userNotificationCenter(_ center: UNUserNotificationCenter,
+                                didReceive response: UNNotificationResponse,
+                                withCompletionHandler completionHandler: @escaping () -> Void) {
```
Comment (MartinJibber) [inline Shared/Notifications/UserNotificationManager.swift:174]: Using non async version of this function so it's called on the main thread.

File: Shared/Extensions/Task+Extensions.swift (hunk 4 above)
Comment (Benjamindodgson) [reply Shared/Extensions/Task+Extensions.swift:25]: So many @'s

File: Shared/Notifications/UserNotificationManager.swift [hunk 10]
```diff
@@ -105,7 +133,7 @@ class UserNotificationManager: NSObject {
                     if note.request.identifier == message.id {
                         identifiers.append(message.id)
                     }
-                    
+
                     if message.context == .timeSensitive {
                         badgeCount -= 1
```
Comment (Benjamindodgson) [inline Shared/Notifications/UserNotificationManager.swift:138]: I forgot that the installation has a badge count on it... refactor for another time.

**Review by Benjamindodgson — APPROVED** [review]
Man what a doozy to get notifications to work. Nice job

**Pull request merged**